RunFlowML does not work when DHF app servers require SSL

When RunFlowML creates its connections to the DHF app servers, it is using the `SSLService` and `SSLHostnameVerifier` from the user's SSL service in NiFi, but it is not carrying over the `X509TrustManager`. Because of this, we run into this error:

Caused by: java.lang.RuntimeException: Unable to retrieve flow with name: CurateCustomerJSON; cause: User is not allowed to POST at /data-hub/5/data-services/flow/getFullFlow.sjs
        at com.marklogic.hub.impl.FlowManagerImpl.getFullFlow(FlowManagerImpl.java:156)
        at com.marklogic.hub.flow.impl.FlowRunnerImpl.runFlow(FlowRunnerImpl.java:179)
        at org.apache.nifi.marklogic.processor.RunFlowMarkLogic.onTrigger(RunFlowMarkLogic.java:153)

There's a separate issue here to analyze as to why we get a message like this when the root cause is due to an SSL misconfiguration. I'll follow up on that in the Java Client project with a test case. 

For the NiFi connector, we need to carry over either the user's provided trust store or a "trust anything" trust store.

--- a/nifi-marklogic-processors/src/main/java/org/apache/nifi/marklogic/processor/RunFlowMarkLogic.java
+++ b/nifi-marklogic-processors/src/main/java/org/apache/nifi/marklogic/processor/RunFlowMarkLogic.java
@@ -4,6 +4,7 @@ import com.fasterxml.jackson.core.type.TypeReference;
 import com.fasterxml.jackson.databind.ObjectMapper;
 import com.marklogic.client.DatabaseClientFactory;
 import com.marklogic.client.ext.DatabaseClientConfig;
+import com.marklogic.client.ext.modulesloader.ssl.SimpleX509TrustManager;
 import com.marklogic.hub.DatabaseKind;
 import com.marklogic.hub.flow.FlowInputs;
 import com.marklogic.hub.flow.FlowRunner;
@@ -25,6 +26,7 @@ import org.apache.nifi.processor.util.StandardValidators;
 import org.apache.nifi.util.StringUtils;
 
 import javax.net.ssl.SSLContext;
+import javax.net.ssl.X509TrustManager;
 import java.io.IOException;
 import java.util.*;
 
@@ -191,6 +193,15 @@ public class RunFlowMarkLogic extends AbstractMarkLogicProcessor {
 				hubConfig.setSslHostnameVerifier(DatabaseKind.FINAL, verifier);
 				hubConfig.setSslHostnameVerifier(DatabaseKind.JOB, verifier);
 			}
+
+			X509TrustManager trustManager = clientConfig.getTrustManager();
+			if (trustManager == null) {
+				getLogger().info("No X509TrustManager found; using 'trust everything' X509TrustManager implementation");
+				trustManager = new SimpleX509TrustManager();
+			}
+			hubConfig.setTrustManager(DatabaseKind.STAGING, trustManager);
+			hubConfig.setTrustManager(DatabaseKind.FINAL, trustManager);
+			hubConfig.setTrustManager(DatabaseKind.JOB, trustManager);
 		}
 
 		String externalName = clientConfig.getExternalName();

--- a/nifi-marklogic-processors/src/test/java/org/apache/nifi/marklogic/processor/RunFlowMarkLogicTest.java
+++ b/nifi-marklogic-processors/src/test/java/org/apache/nifi/marklogic/processor/RunFlowMarkLogicTest.java
@@ -12,6 +12,8 @@ import com.marklogic.hub.impl.HubConfigImpl;
 import org.junit.Before;
 import org.junit.Test;
 
+import javax.net.ssl.SSLContext;
+import javax.net.ssl.X509TrustManager;
 import java.util.Map;
 
 public class RunFlowMarkLogicTest extends AbstractMarkLogicProcessorTest {
@@ -93,5 +95,53 @@ public class RunFlowMarkLogicTest extends AbstractMarkLogicProcessorTest {
 		assertNull(inputs.getSteps());
 		assertNull(inputs.getJobId());
 		assertNull(inputs.getOptions());
+	}
+
+	@Test
+	public void configureSslWithNoTrustManager() {
+		DatabaseClientConfig config = new DatabaseClientConfig("somehost", 9010, "someuser", "someword");
+		final SSLContext sslContext = SimpleX509TrustManager.newSSLContext();
+		config.setSslContext(sslContext);
+		config.setSslHostnameVerifier(DatabaseClientFactory.SSLHostnameVerifier.STRICT);
+		config.setTrustManager(null);
+
+		HubConfigImpl hubConfig = processor.initializeHubConfig(processContext, config);
+
+		assertTrue(sslContext == hubConfig.getSslContext(DatabaseKind.STAGING));
+		assertTrue(sslContext == hubConfig.getSslContext(DatabaseKind.FINAL));
+		assertTrue(sslContext == hubConfig.getSslContext(DatabaseKind.JOB));
+		assertEquals(DatabaseClientFactory.SSLHostnameVerifier.STRICT, hubConfig.getSslHostnameVerifier(DatabaseKind.STAGING));
+		assertEquals(DatabaseClientFactory.SSLHostnameVerifier.STRICT, hubConfig.getSslHostnameVerifier(DatabaseKind.FINAL));
+		assertEquals(DatabaseClientFactory.SSLHostnameVerifier.STRICT, hubConfig.getSslHostnameVerifier(DatabaseKind.JOB));
+
+		String message = "Since no trust manager was provided, a simple trust-everything one should be used";
+		assertTrue(message, hubConfig.getTrustManager(DatabaseKind.STAGING) instanceof SimpleX509TrustManager);
+		assertTrue(message, hubConfig.getTrustManager(DatabaseKind.FINAL) instanceof SimpleX509TrustManager);
+		assertTrue(message, hubConfig.getTrustManager(DatabaseKind.JOB) instanceof SimpleX509TrustManager);
+	}
+
+	@Test
+	public void configureSslWithTrustManager() {
+		DatabaseClientConfig config = new DatabaseClientConfig("somehost", 9010, "someuser", "someword");
+		final SSLContext sslContext = SimpleX509TrustManager.newSSLContext();
+		final X509TrustManager trustManager = new SimpleX509TrustManager();
+		config.setSslContext(sslContext);
+		config.setSslHostnameVerifier(DatabaseClientFactory.SSLHostnameVerifier.COMMON);
+		config.setTrustManager(trustManager);
+
+		HubConfigImpl hubConfig = processor.initializeHubConfig(processContext, config);
+
+		assertTrue(sslContext == hubConfig.getSslContext(DatabaseKind.STAGING));
+		assertTrue(sslContext == hubConfig.getSslContext(DatabaseKind.FINAL));
+		assertTrue(sslContext == hubConfig.getSslContext(DatabaseKind.JOB));
+		assertEquals(DatabaseClientFactory.SSLHostnameVerifier.COMMON, hubConfig.getSslHostnameVerifier(DatabaseKind.STAGING));
+		assertEquals(DatabaseClientFactory.SSLHostnameVerifier.COMMON, hubConfig.getSslHostnameVerifier(DatabaseKind.FINAL));
+		assertEquals(DatabaseClientFactory.SSLHostnameVerifier.COMMON, hubConfig.getSslHostnameVerifier(DatabaseKind.JOB));
+
+		String message = "Should use the trust manager in the DatabaseClientConfig, which will have been configured " +
+				"by the user in the NiFi SSL service";
+        assertTrue(message, trustManager == hubConfig.getTrustManager(DatabaseKind.STAGING));
+        assertTrue(message, trustManager == hubConfig.getTrustManager(DatabaseKind.FINAL));
+        assertTrue(message, trustManager == hubConfig.getTrustManager(DatabaseKind.JOB));
 	}
 }

--- a/nifi-marklogic-services/src/main/java/org/apache/nifi/marklogic/controller/DefaultMarkLogicDatabaseClientService.java
+++ b/nifi-marklogic-services/src/main/java/org/apache/nifi/marklogic/controller/DefaultMarkLogicDatabaseClientService.java
@@ -33,6 +33,7 @@ import org.apache.nifi.controller.ConfigurationContext;
 import org.apache.nifi.expression.ExpressionLanguageScope;
 import org.apache.nifi.processor.exception.ProcessException;
 import org.apache.nifi.processor.util.StandardValidators;
+import org.apache.nifi.security.util.ClientAuth;
 import org.apache.nifi.security.util.KeyStoreUtils;
 import org.apache.nifi.ssl.SSLContextService;
 
@@ -144,7 +145,7 @@ public class DefaultMarkLogicDatabaseClientService extends AbstractControllerSer
         .description("Client authentication policy when connecting via a secure connection. This property is only used when an SSL Context "
             + "has been defined and enabled")
         .required(false)
-        .allowableValues(SSLContextService.ClientAuth.values())
+        .allowableValues(ClientAuth.values())
         .expressionLanguageSupported(ExpressionLanguageScope.VARIABLE_REGISTRY)
         .build();
 
@@ -199,14 +200,12 @@ public class DefaultMarkLogicDatabaseClientService extends AbstractControllerSer
 
         final SSLContextService sslService = context.getProperty(SSL_CONTEXT_SERVICE).asControllerService(SSLContextService.class);
         if(sslService != null) {
-            final SSLContextService.ClientAuth clientAuth = determineClientAuth(context);
+            final ClientAuth clientAuth = determineClientAuth(context);
+            getLogger().info("Configuring SSL connection; client authentication: " + clientAuth);
             try {
                 if (sslService.isTrustStoreConfigured()) {
+                    getLogger().info("Configuring TrustManager based on trust store found in SSLService");
                     final TrustManagerFactory trustManagerFactory = TrustManagerFactory.getInstance(TrustManagerFactory.getDefaultAlgorithm());
-//                    final KeyStore trustStore = KeyStoreUtils.getTrustStore(sslService.getTrustStoreType());
-//                    try (final InputStream is = new FileInputStream(sslService.getTrustStoreFile())) {
-//                        trustStore.load(is, );
-//                    }
                     final KeyStore trustStore = KeyStoreUtils.loadTrustStore(
                             sslService.getTrustStoreFile(),
                             sslService.getTrustStorePassword().toCharArray(),
@@ -227,13 +226,13 @@ public class DefaultMarkLogicDatabaseClientService extends AbstractControllerSer
         return config;
     }
 
-    protected SSLContextService.ClientAuth determineClientAuth(ConfigurationContext context) {
+    protected ClientAuth determineClientAuth(ConfigurationContext context) {
         try {
-            return context.getProperty(CLIENT_AUTH).getValue() == null ? SSLContextService.ClientAuth.REQUIRED :
-                    SSLContextService.ClientAuth.valueOf(context.getProperty(CLIENT_AUTH).evaluateAttributeExpressions().getValue());
+            return context.getProperty(CLIENT_AUTH).getValue() == null ? ClientAuth.REQUIRED :
+                    ClientAuth.valueOf(context.getProperty(CLIENT_AUTH).evaluateAttributeExpressions().getValue());
         } catch (IllegalArgumentException exception) {
             throw new ProviderCreationException("Client Authentication should be one of the following values : "
-                + Arrays.toString(SSLContextService.ClientAuth.values()));
+                + Arrays.toString(ClientAuth.values()));
         }
     }
 

--- a/nifi-marklogic-services/src/test/java/org/apache/nifi/marklogic/controller/EvaluateExpressionsTest.java
+++ b/nifi-marklogic-services/src/test/java/org/apache/nifi/marklogic/controller/EvaluateExpressionsTest.java
@@ -5,6 +5,7 @@ import com.marklogic.client.ext.SecurityContextType;
 import org.apache.nifi.components.PropertyDescriptor;
 import org.apache.nifi.expression.ExpressionLanguageScope;
 import org.apache.nifi.registry.VariableDescriptor;
+import org.apache.nifi.security.util.ClientAuth;
 import org.apache.nifi.ssl.SSLContextService;
 import org.apache.nifi.util.MockConfigurationContext;
 import org.apache.nifi.util.MockVariableRegistry;
@@ -99,9 +100,9 @@ public class EvaluateExpressionsTest extends Assert {
 	@Test
 	public void evaluateClientAuth() {
 		verifyScope(DefaultMarkLogicDatabaseClientService.CLIENT_AUTH);
-		variableRegistry.setVariable(new VariableDescriptor("myValue"), SSLContextService.ClientAuth.WANT.name());
+		variableRegistry.setVariable(new VariableDescriptor("myValue"), "WANT");
 		properties.put(DefaultMarkLogicDatabaseClientService.CLIENT_AUTH, "${myValue}");
-		assertEquals(SSLContextService.ClientAuth.WANT, service.determineClientAuth(context));
+		assertEquals(ClientAuth.WANT, service.determineClientAuth(context));
 	}
 
 	private void verifyScope(PropertyDescriptor descriptor) {

